app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageCompose.java

```diff
@@ -363,7 +364,25 @@ public void onTextChanged(CharSequence s, int start, int before, int count) {
 
         subjectView.addTextChangedListener(draftNeedsChangingTextWatcher);
 
-        messageContentView.addTextChangedListener(draftNeedsChangingTextWatcher);
+        TextWatcher messageContentViewTextWatcher = new SimpleTextWatcher(){
+            @Override
+            public void onTextChanged(CharSequence s, int start, int before, int count) {
+                changesMadeSinceLastSave = true;
+                int smallestSchemaSize = 5;
+                int addedCharactersLength = 2;
+                if (count >= smallestSchemaSize) {
+                    CharSequence pastedText = s.subSequence(start, start + count);
+                    if (UriMatcher.INSTANCE.isValidUri(pastedText)) {
+                        String newText = addAngleBracketsToPastedUri(s, start, count, pastedText);
+                        messageContentView.removeTextChangedListener(this);
+                        messageContentView.setText(newText);
+                        messageContentView.setSelection(start + count + addedCharactersLength);
+                        messageContentView.addTextChangedListener(this);
+                    }
+                }
+            }
+        };
```

Comment:
In `TextWatcher.afterTextChanged()` the text can be changed. So this whole feature can be written as an independent `TextWatcher` implementation, i.e. one without any dependency on `MessageCompose`.